File: c++/0725.cpp
// Link: https://leetcode.com/problems/split-linked-list-in-parts/description/
// Runtime: 9 ms
// Memory: 8.92 MB

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution 
{
public:
    std::vector<ListNode*> splitListToParts(ListNode* head, int k) 
    {
        ListNode* headToCountLength = head;
        int lengthOfList = 0;
        while (headToCountLength != nullptr)
        {
            lengthOfList++;
            headToCountLength = headToCountLength->next;
        }

        std::vector<ListNode*> partitionedList(k);
        ListNode* iterator = head;
        int lengthOfPartition = lengthOfList / k, remainder = lengthOfList % k;
        for (int i = 0; i < k; i++)
        {
            ListNode* currentHead = iterator;
            int j = 0;
            while ((j < lengthOfPartition + (i < remainder) - 1) && (iterator != nullptr))
            {
                iterator = iterator->next;                
                j++;
            }

            if (iterator != nullptr)
            {
                ListNode* previous = iterator;
                iterator = iterator->next;
                previous->next = nullptr;
            }

            partitionedList[i] = currentHead;
        }

        return partitionedList;
    }
};

File: python/0725.py
# Link: https://leetcode.com/problems/split-linked-list-in-parts/description/
# Runtime: 35 ms
# Memory: 16.71 MB

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        head_to_count_length, length_of_list = head, 0
        while head_to_count_length != None:
            length_of_list += 1
            head_to_count_length = head_to_count_length.next
        
        partitioned_list = [None] * k
        length_of_partition, remainder = length_of_list // k, length_of_list % k
        iterator = head
        for i in range(k):
            current_head = iterator
            j = 0
            while (j < length_of_partition + (i < remainder) - 1) and (current_head != None):
                iterator = iterator.next
                j += 1
            
            if iterator != None:
                previous = iterator
                iterator = iterator.next
                previous.next = None
            
            partitioned_list[i] = current_head
        
        return partitioned_list
        